### ISSUE TITLE
Azure extension is impractical when bootstrapping servers with lengthy run lists

### ISSUE THREAD

**Issue by juliandunn** [chef-partners/azure-chef-extension#9]
I have found the new Azure client extension an interesting though ultimately frustrating way of bootstrapping Azure VMs, because it looks like Microsoft's agent enforces a 5 minute timeout for any agent plugin to complete. After this timeout, it kills the plugin:

```
[00000004] [01/20/2015 02:54:32.29] [INFO]  Staring a process with the launch command C:\Packages\Plugins\Chef.Bootstrap.WindowsAzure.ChefClient\11.16.4.2\enable.cmd and params:
[00000004] [01/20/2015 02:54:32.29] [INFO]  Beginning enable of plugin Chef.Bootstrap.WindowsAzure.ChefClient.
[00000004] [01/20/2015 02:54:32.29] [INFO]  Waiting for command C:\Packages\Plugins\Chef.Bootstrap.WindowsAzure.ChefClient\11.16.4.2\enable.cmd of plugin (name Chef.Bootstrap.WindowsAzure.ChefClient version 11.16.4.2) to finish...
[00000004] [01/20/2015 02:59:35.32] [ERROR] Command C:\Packages\Plugins\Chef.Bootstrap.WindowsAzure.ChefClient\11.16.4.2\enable.cmd of Chef.Bootstrap.WindowsAzure.ChefClient has not exited on time! Killing it...
[00000004] [01/20/2015 02:59:35.32] [ERROR] Enable command of Chef.Bootstrap.WindowsAzure.ChefClient has exited with Exit code: -2
[00000004] [01/20/2015 02:59:35.33] [WARN]  Enable failed for plugin (name: Chef.Bootstrap.WindowsAzure.ChefClient, version 11.16.4.2) with exception Command C:\Packages\Plugins\Chef.Bootstrap.WindowsAzure.ChefClient\11.16.4.2\enable.cmd of
 Chef.Bootstrap.WindowsAzure.ChefClient has not exited on time! Killing it..., Code: 1009
[00000004] [01/20/2015 02:59:35.47] [ERROR] Error while processing plugin Chef.Bootstrap.WindowsAzure.ChefClient version 11.16.4.2, state: enabled, error: System.Exception: Command C:\Packages\Plugins\Chef.Bootstrap.WindowsAzure.ChefClient\11.16.4.2\enable.cmd of Chef.Bootstrap.WindowsAzure.ChefClient has not exited on time! Killing it...
   at Microsoft.WindowsAzure.GuestAgent.ContainerStateMachine.HandlerStateMachine.InvokeCommand(String command, PluginArtifacts pluginArtifact, String pluginVersion, String pluginFolder, String pluginLogFolder, Int32 processWaitTimeout, PluginEventType startType, PluginEventType endType)
   at Microsoft.WindowsAzure.GuestAgent.ContainerStateMachine.HandlerStateMachine.EnableHandler(PluginArtifacts artifact)
   at Microsoft.WindowsAzure.GuestAgent.ContainerStateMachine.HandlerStateMachine.StartHandlerStateMachine(PluginArtifacts data, RDConfigHostingEnvironmentSettingsPlugin plugin)
   at Microsoft.WindowsAzure.GuestAgent.ContainerStateMachine.PluginInstaller.PluginBackgroundThread(Object sender, DoWorkEventArgs e)
```

This "feature" makes the bootstrapping of servers with anything more than a trivial run_list to be totally impractical because it is common for an initial Chef run to last more than 5 minutes.

Could we talk with MSFT about this limit?

**Comment by adamedx** [comment]
@NimishaS this should no longer be an issue, correct? We no longer perform bootstrap within enable command, right?

**Comment by NimishaS** [comment]
@adamedx , we do perform bootstrap within enable command but timeout error doesn't occur anymore. We tested this by creating a vm with extension having a recipe which takes more than 5 mins to run and it worked fine.

**Comment by cparedes** [comment]
I am having this issue as well - I have one recipe in my run_list, with an expanded run_list of about seven recipes. This seems to bomb out consistently even if Chef successfully completes successfully via running chef-client by hand. This is with knife-azure 1.4.0 and master on knife-windows (1.0.0.dev.0)

**Comment by adamedx** [comment]
@cparedes, which version of Chef are you using? The latest extension versions should have addressed this problem. Are you using the portal to start the extension, or the knife-azure / azure-xplat / azure powershell command line tools?

**Comment by cparedes** [comment]
This is on Chef 12.3. I'm using knife-azure.

**Comment by cparedes** [comment]
Is there a way to figure out which versions of the extension are available? I believe it's possible to specify the exact extension version too, right?

**Comment by NimishaS** [comment]
Hello @cparedes, it's possible to specify the exact extension version in `knife.rb` file as `knife[:azure_chef_extension_version] = '1207.12'`. `1207.12.3.0` version should fix your problem.

For getting the latest available extension version, use this command on azure powershell CLI
```
Get-AzureVMAvailableExtension  | select ExtensionName,Publisher,Version,PublishedDate
```

**Comment by cparedes** [comment]
Awesome, will try this today and will report back my results.

**Comment by cparedes** [comment]
Oh wow, yeah that worked perfectly - things complete within 2 minutes even with an expanded run_list that'd probably run over the 5 minute limit. Thanks a ton!

**Comment by adamedx** [comment]
Thank you for all your patience @cparedes, and I'm glad you were able to get past this.

**Comment by tekmaven** [comment]
I am still having this issue, including when I specify the exact version listed above:

![image](https://cloud.githubusercontent.com/assets/349445/8509477/b48c96ec-2273-11e5-96ed-0a2c54c7eb29.png)

**Comment by cparedes** [comment]
I'm getting the issue again as well with 1207.12. Was working for a while
but its starting to bomb out consistently again.

**Comment by tekmaven** [comment]
@NimishaS Any way to re-open this?

**Comment by kgottumukkala** [comment]
I am also getting same issue. Attached is the screen shot.


Publisher

Chef.Bootstrap.WindowsAzure


Version

1207.12.3.0


Status

NotReady
 

Message

Enable failed for plugin (name: Chef.Bootstrap.WindowsAzure.ChefClient, version 1207.12.3.0) with exception Command C:\Packages\Plugins\Chef.Bootstrap.WindowsAzure.ChefClient\1207.12.3.0\enable.cmd of Chef.Bootstrap.WindowsAzure.ChefClient has not exited on time! Killing it...

Let me know if its fixed.


![image](https://cloud.githubusercontent.com/assets/10495051/8519550/0f5acffc-23f1-11e5-86e5-3519e4e3a07b.png)

**Comment by NimishaS** [comment]
@kgottumukkala , can you share following details:
1. Did you create VM using `knife-azure` or `Azure Powershell cmdlet`
2. What was the VM size?
3. What is the service location?

**Comment by kgottumukkala** [comment]
@NimishaS :+1: 

Thanks for the response.

1. I am using Azure PowerShell Cmdlet
2. VM size: Small
3. Service location : East US

**Comment by NimishaS** [comment]
@kgottumukkala , we have tried with windows 2012R2 and 2012 but haven't been able to reproduce the timeout issue yet. Can you please share logs from the following locations?
1. `C:\WindowsAzure\Logs\WaAppAgent`
2. All file `C:\WindowsAzure\Logs\Plugins\Chef.Bootstrap.WindowsAzure.ChefClient\1207.12.3.0` location

**Comment by tekmaven** [comment]
@NimishaS I think the key to the bug is using an A1 VM.  The chef client install itself takes some time and the A1 specs (and I'm assuming A0 too) do not make it install as fast as possible.

If I do not configure a run list, then the extension reliability installs.

My workaround is to provision with empty run lists and then assign after the extension completes.

**Comment by cparedes** [comment]
I've hit this with a Large instance as well. Based on what's going on right now, we try to fork off another process to do the Chef run to reduce the time it takes to return a 'success' back over to the API. Is there a way to return success first thing during an enable, and if not, is it possible to 'disown' the forked process so it runs after the parent process forks off the Chef run?

**Comment by tekmaven** [comment]
I just created four A1s, no run list and they all failed :-1:

**Comment by TheD0kt0r** [comment]
I just tried provisioning a Basic_A1 in Australia East and Australia Southeast - both VMs failed with the same errors shown above (timeout).  That's without applying a runlist - simply installing the Chef extension and having it register with the Chef server.

Tried again with a Standard_A2 VM in Australia Southeast, and this time it completed successfully.

One thing I have noticed is that every VM I provision with the Chef extension is installing an earlier version of the client: 11.18.6.1.  Is this an Azure regional issue?

**Comment by tekmaven** [comment]
@TheD0kt0r Are you using portal.azure.com to provision? Or knife? Or azure powershell?

I find that if I use portal.azure.com, I get 11.x.  If I use Azure Powershell, Set-AzureVMChefExtension will use the latest 12.x version.

**Comment by TheD0kt0r** [comment]
@tekmaven Yes, I'm using the Portal. That's interesting feedback. Strange that the portal uses one version whereas the tools use a different one.  Thanks though - I'll change my demo around to use PowerShell.  It would also be worth checking whether the new Preview Portal shows the same behaviour....

**Comment by NimishaS** [comment]
We are looking into this issue.
@tekmaven, there has been a change in the version scheme of `azure-chef-extension` where we have 4 digit major version number now(Please refer https://github.com/chef-partners/azure-chef-extension#azure-chef-extension-version-scheme). Microsoft Azure Portal is not supporting this currently. They will be fixing it soon.

**Comment by NimishaS** [comment]
Issue we are facing currently is not run-list related. Hence closing this issue. Created a new issue https://github.com/chef-partners/azure-chef-extension/issues/47

**Comment by NimishaS** [comment]
@tekmaven , @TheD0kt0r , @cparedes , @kgottumukkala , have you faced this issue on windows only or linux too?

**Comment by kgottumukkala** [comment]
On Windows.. However if I take VM size as Basic_A2 or more, its not throwing any issue, Size of the Ram 1.75Gb in Basic_A1 might be an issue as @tekmaven pointed.

**Comment by TheD0kt0r** [comment]
There's no option at the moment to enable the Chef extension on a Linux VM via the Azure portal, so I've only tested this with Windows.

**Comment by NimishaS** [comment]
@TheD0kt0r , issue is resolved with `azure-chef-extension 1210.12.4.1`
